audit/entry_formatter.go

```diff
@@ -623,3 +638,71 @@ func newTemporaryEntryFormatter(n *EntryFormatter) *EntryFormatter {
 func (s *nonPersistentSalt) Salt(_ context.Context) (*salt.Salt, error) {
 	return salt.NewNonpersistentSalt(), nil
 }
+
+// excludeFields takes an entry (RequestEntry or ResponseEntry) and attempts to
+// exclude fields that have been configured on the formatter node.
+func (f *EntryFormatter) excludeFields(entry any) (map[string]any, error) {
+	m := make(map[string]any)
+	err := mapstructure.Decode(entry, &m)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding entry: %v", err)
+	}
+
+	// If there are no exclusions to process then return the map.
+	if len(f.exclusions) < 1 {
```

Comment:
You're right and right to point it out 😄 .

The reason I ended up doing it this way was to make sure that the ordering looked the same in the audit log regardless of whether there were exclusions or not.

Reply:
Hmm, could you go into detail about what the ordering difference might be? I'm very cautious about increasing performance overhead for auditing so I want to properly understand the implications of what we might be gaining by adding this for all requests.

Reply:
When we decode into a map, then eventually back to JSON as bytes, the map is ordered with A-Z keys. 

So the bytes as they're written out then are in that order, if we only decode some things and not others you will get an audit log where the first value in the JSON isn't the same between log lines.

`{"time":"2024-01-...`
`{"auth":{"access...`

The filtering and exclusion features both use a lot of reflection, so I don't imagine we're really going to make a big difference with this 'into a map' stuff. (But if the benchmarking does start showing issues then we've got places to come looking).

Reply:
Ahhh I see! Thanks for the explanation. This will be a change upon upgrade to the version of Vault with this exclusion in, then, right? i.e. my old log file will start getting differently ordered logs in it.

It definitely makes sense to keep these the same, but I do feel wary about the impact. If we're already planning to benchmark it then I feel a lot more comfortable.

>The filtering and exclusion features both use a lot of reflection, so I don't imagine we're really going to make a big difference with this 'into a map' stuff. 

Yeah, that makes sense, my worry is more that this will affect users who aren't using either of those features. It's probably only a small increase, but it is a small increase for every request! If I don't want to use this feature and my auditing is 5% slower, that's 5% less requests Vault can handle.

It could be that we'd have one 'ordering' for Vault implementations with exclusions defined, and another for those that don't have any, for example. It's a little less 'elegant' but it's more performant. It's an option and worth thinking about!

Still, I might be putting the cart before the horse here, if we're planning on benchmarking anyway it could be that this has only a 0.1% impact which is close to meaningless.

Reply:
My fears are mostly allayed but it was really useful to talk through this -- I do want to keep this thread open though to see if others have thoughts on this trade-off!

Reply:
Problem solved (see the Jira for benchmarks), thanks for the discussion and bringing attention to it early on. 👍🏼